### PR TITLE
conv3d_ndhwc schedule

### PR DESCRIPTION
**Pull request by alexgl-github**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

### CODE REVIEW

File: topi/python/topi/x86/conv3d.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        # inline all one-to-one-mapping operators except the last stage (output)
+        if tag.is_broadcast(op.tag):
+            if op not in s.outputs:
+                s[op].compute_inline()
+            else: # inject custom schedule
+                if len(op.axis) == 5: # schedule bias + bn + relu
+                    n, d, h, w, c = op.axis
+                    fused = s[op].fuse(n, d, h, w)
+                    s[op].parallel(fused)
+                    s[op].vectorize(c)
```
Comment (anijain2305) [inline topi/python/topi/x86/conv3d.py]: Is this applicable?

File: topi/python/topi/x86/conv3d.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                kxi, kxo = s[C].split(kx, factor=2)
+                kci, kco = s[C].split(kc, factor=2)
+                #
+                s[C].reorder(n, d, ho, wo, hi, wi, c, kxo, kco, kxi, kci)
+                s[C].unroll(kci)
+
+            s[C].vectorize(c)
+            if op != output_op:
+                _, _, _, _, c_out = output_op.axis
+                s[C].compute_at(s[output_op], c_out)
```
Comment (anijain2305) [inline topi/python/topi/x86/conv3d.py]: Another lines to think more about - I think compute_at might be better at h axis.

File: topi/python/topi/x86/conv3d.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+@autotvm.register_topi_compute(conv3d, 'cpu', ['direct'])
+def conv3d_x86(cfg, input, filter, strides, padding, dilation, layout='NCDHW', out_dtype=None):
+    if layout == 'NCDHW':
+        return conv3d_ncdhw(input, filter, strides, padding, dilation, out_dtype)
+    elif layout == 'NDHWC':
+        return conv3d_ndhwc(input, filter, strides, padding, dilation, out_dtype)
+
+@autotvm.register_topi_schedule(schedule_conv3d_ndhwc, 'cpu', ['direct'])
+def schedule_conv3d_ndhwc_x86(cfg, outs):
+    """TOPI schedule callback for conv2d
```
Comment (apivovarov) [inline topi/python/topi/x86/conv3d.py]: 3d

**Comment by alexgl-github** [comment]
@apivovarov 
@anijain2305 
Could you please review again?

File: topi/python/topi/x86/conv3d.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                data_pad = data
+                data = data_pad.op.input_tensors[0]
+                n_pad, d_pad, h_pad, w_pad, c_pad = data_pad.op.axis
+                pad_fused = s[data_pad].fuse(h_pad, w_pad)
+                s[data_pad].parallel(pad_fused)
+            C = conv
+            n, d, h, w, c = s[C].op.axis
+            s[C].vectorize(c)
+            if op != output_op:
+                n_out, d_out, h_out, w_out, c_out = output_op.axis
```
Comment (anijain2305) [inline topi/python/topi/x86/conv3d.py]: Suggest to use _ for unused axes.

File: topi/python/topi/x86/conv3d.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+            data = op.input_tensors[0]
+            data_pad = None
+            if isinstance(data.op, tvm.tensor.ComputeOp) and "pad" in data.op.tag:
+                data_pad = data
+                data = data_pad.op.input_tensors[0]
+                n_pad, d_pad, h_pad, w_pad, c_pad = data_pad.op.axis
+                pad_fused = s[data_pad].fuse(h_pad, w_pad)
+                s[data_pad].parallel(pad_fused)
+            C = conv
```
Comment (anijain2305) [inline topi/python/topi/x86/conv3d.py]: Some comments will be helpful here and there to pick it up later.

File: topi/python/topi/x86/conv3d.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            kernel = op.input_tensors[1]
+            if isinstance(kernel.op, tvm.tensor.ComputeOp) and "dilate" in kernel.op.tag:
+                s[kernel].compute_inline()
+
+            data = op.input_tensors[0]
+            data_pad = None
+            if isinstance(data.op, tvm.tensor.ComputeOp) and "pad" in data.op.tag:
+                data_pad = data
+                data = data_pad.op.input_tensors[0]
+                n_pad, d_pad, h_pad, w_pad, c_pad = data_pad.op.axis
```
Comment (anijain2305) [inline topi/python/topi/x86/conv3d.py]: Suggest to use _ for unused axes.

**Comment by anijain2305** [comment]
@vinx13 @kevinthesun  for further reviews.

**Comment by alexgl-github** [comment]
> @alexgl-github Thanks for the contribution.
> 
> Almost every PR needs to have a test case in Apache TVM project. In this case, it can be a test case that uses this schedule and compares accuracy with some golden reference.

There's already topi conv3d_ndhwc test using this schedule:
https://github.com/apache/incubator-tvm/blob/1b8522e475e9a889b80f069a83928cafa3502a74/topi/tests/python/test_topi_conv3d_ndhwc.py#L60

File: topi/python/topi/x86/conv3d.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import tvm
+from .. import generic, tag
+
+@generic.schedule_conv3d_ndhwc.register("cpu")
+def schedule_conv3d_ndhwc(outs):
+    """TOPI schedule callback for conv2d
+
+    Parameters
+    ----------
+    cfg: ConfigEntity
```
Comment (vinx13) [inline topi/python/topi/x86/conv3d.py]: this argument does not exist

File: topi/python/topi/x86/conv3d.py (hunk 3 above)
Comment (vinx13) [reply topi/python/topi/x86/conv3d.py]: @alexgl-github this comment hasn't been resolved

Comment (alexgl-github) [reply topi/python/topi/x86/conv3d.py]: @vinx13 Fixed, thanks.

File: topi/python/topi/x86/conv3d.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                _, _, _, _, c_out = output_op.axis
+                s[C].compute_at(s[output_op], c_out)
+            else:
+                # fuse batch, depth, height axes
+                fused = s[C].fuse(n, d, h)
+                s[C].parallel(fused)
+
+        scheduled_ops.append(op)
+
+    traverse(output_op)
```
Comment (icemelon9) [inline topi/python/topi/x86/conv3d.py]: You can use traverse_inline function to avoid redundant inlining broadcast ops.

Comment (alexgl-github) [reply topi/python/topi/x86/conv3d.py]: @icemelon9 Changed to traverse_inline, thanks

File: topi/python/topi/x86/conv3d.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        in the format of an array of tensors.
+
+    Returns
+    -------
+    s: Schedule
+        The computation schedule for conv3d.
+    """
+    s = tvm.create_schedule([x.op for x in outs])
+    output_op = outs[0].op
+    scheduled_ops = []
```
Comment (icemelon9) [inline topi/python/topi/x86/conv3d.py]: Remove this line since it's useless now

File: topi/python/topi/x86/conv3d.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            s[C].vectorize(c)
+            if op != output_op: # fuse bias + bn + activation
+                _, _, _, _, c_out = output_op.axis
+                s[C].compute_at(s[output_op], c_out)
+            else:
+                # fuse batch, depth, height axes
+                fused = s[C].fuse(n, d, h)
+                s[C].parallel(fused)
+
+        scheduled_ops.append(op)
```
Comment (icemelon9) [inline topi/python/topi/x86/conv3d.py]: remove this line

**Comment by icemelon9** [comment]
Thanks @alexgl-github . This is now merged.

**Pull request merged**